hogben/utils.py

```diff
@@ -163,77 +157,182 @@ def fisher(qs: list[np.ndarray],
+                 models: list[Union['refnx.reflect.ReflectModel',
+                                    'refl1d.experiment.Experiment']],
+                 step: float = 0.005):
+        """Constructor for the Fisher matrix class.
+
+        Args:
+            qs: The Q points for each model.
+            xi: The varying model parameters.
+            counts: incident neutron counts corresponding to each Q value.
+            models: models to calculate gradients with.
+            step: step size to take when calculating gradient.
+            g: The Fisher information matrix
+            min_eigenval: The minimum eigenvalue associated with the Fisher
+             information matrix
+        """
+        self.qs = qs
+        self.xi = xi
+        self.counts = counts
+        self.models = models
+        self.step = step
+
+    @property
+    def n(self) -> int:
+        """The total amount of datapoints.
+
+        Returns:
+            int: total amount of datapoints.
+        """
+        return sum(len(q) for q in self.qs)
+
+    @property
+    def m(self) -> int:
+        """The total amount of parameters.
+
+        Returns:
+            int: total amount of parameters.
+        """
+        return len(self.xi)
+
+    @property
+    def g(self) -> np.ndarray:
+        """The Fisher information matrix.
+
+        Returns:
+            numpy.ndarray: The Fisher information matrix.
+        """
+        return self._calculate_g()
+
+    @g.setter
+    def g(self, value: np.ndarray):
+        """Sets the fisher information matrix."""
+        self.g = value
+
+    @property
+    def min_eigenval(self) -> float:
+        """The minimum eigenvalue associated with the Fisher information
+         matrix.
+
+        Returns:
+            float: The minimum eigenvalue.
+        """
+        return np.linalg.eigvalsh(self.g)[0]
+
+    def _calculate_g(self) -> np.ndarray:
+        """Calculates the Fisher information matrix using the class attributes.
+
+        Returns:
+            numpy.ndarray: The Fisher information matrix.
+        """
+        if self.n == 0:
+            return np.zeros((self.m, self.m))
+        J = self._get_gradient_matrix()
+
+        # Calculate the reflectance for each model for the given Q values.
+        r = np.concatenate([reflectivity(q, model)
+                            for q, model in list(zip(self.qs, self.models))])
+
+        # Calculate the Fisher information matrix using equations from
+        # the paper.
+        M = np.diag(np.concatenate(self.counts) / r, k=0)
+        g = np.dot(np.dot(J.T, M), J)
+
+        # Perform unit scaling if there's more than one parameter
+        if len(self.xi) > 1:
+            g = self._scale_units(g)  # Scale by unit bounds
+            g = self._scale_importance(g)  # Scale by importance
+        return g
+
+    def _scale_units(self, g: np.ndarray) -> np.ndarray:
+        """Scale the Fisher information matrix using unit scaling.
```

Comment:
"Scale the values of the fisher information matrix for each parameter from interval [lb, ub] to the interval [0, 1]